### PR TITLE
[Event Hubs Client] Custom Endpoint Support

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to add an option for specifying a custom address to use for establishing the connection to the Event Hubs service; the namespace address will continue to be used for communication with the service once the connection is established when the custom endpoint is configured.

**Note:** These changes have discussed with, and approved, the .NET language architect. Marking with the corresponding tag for tracking purposes.

# Last Upstream Rebase

Friday, January 15, 2:56pm (EST)

# References and Related

- [Board Review: Event Hubs - Custom Endpoint Support (#1367)](https://github.com/Azure/azure-sdk/issues/1367)
- [Design: Event Hubs Custom Endpoint Support](https://gist.github.com/jsquire/f0948c9a5ffe2741b03d312dcf8d899e)
- [Custom Endpoint Address Support (#12555)](https://github.com/Azure/azure-sdk-for-net/issues/136712555)

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpClient.cs [hunk 1]
```diff
@@ -233,7 +250,7 @@ internal class AmqpClient : TransportClient
                         ++failedAttemptCount;
                         retryDelay = retryPolicy.CalculateRetryDelay(activeEx, failedAttemptCount);
 
-                        if ((retryDelay.HasValue) && (!ConnectionScope.IsDisposed) && (!cancellationToken.IsCancellationRequested))
+                        if ((retryDelay.HasValue) && (!ConnectionScope.IsDisposed) && (!_closed) && (!cancellationToken.IsCancellationRequested))
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpClient.cs:253]: This change seems unrelated to the custom endpoint. Is there a test covering it?

**Review by christothes — APPROVED** [review]

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpClient.cs (hunk 1 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpClient.cs:253]: They're orthogonal, but were overlooked earlier.   Unfortunately, we're deep enough in the AMQP space at this point that we can't unit test due to the lack of mocking support.    

In theory, this is redundant to the connection scope check, but there may be a slight delay for network operations there causing a (mostly) benign race.  The `_closed` check should not change the behavior but should give us a more deterministic detection.
